Added the transcript for the talk. (#783)
nf-core/nf-co.re: markdown/events/2021/bytesize-12-template-sync.md

* YouTube: <https://youtu.be/-CZKoo5Y_08>
* Bilibili: <https://www.bilibili.com/video/BV1D64y127yJ/>
* FigShare: <https://doi.org/10.6084/m9.figshare.14572866.v1>

<details markdown="1"><summary>Video transcription</summary>
**Note: The content has been edited to make it reader-friendly**

> [0:46](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=46) Hi everyone, I’ll be going over how template synchronization works. This talk is targeted at pipeline developers within the nf core ecosystem and for those who are interested in building their own pipeline based on the nf-core toolsets.

Reviewer: ```suggestion
[0:46](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=46) Hi everyone, I’ll be going over how template synchronization works. This talk is targeted at pipeline developers within the nf-core ecosystem and for those who are interested in building their own pipeline based on the nf-core toolsets.
```

> [1:03](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=63) First, a refresher here. All of this is based very heavily on Git and GitHub, so let’s very quickly go over some of what we’ve covered before [Bytesize#4](https://nf-co.re/events/2021/bytesize-4-github-contribution-basics). Git is a version control code system that has a bit of jargon that you need to know. You do `commit`, switching little packages of work so you do some edits and then bubble them together and then say `commit` and that’s like checkpoints in the timeline of your code. They go in a kind of straight line, and then you can make branches, where you can work on different things at the same time.

Reviewer: ```suggestion
[1:03](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=63) First, a refresher here. All of this is based very heavily on Git and GitHub, so let’s very quickly go over some of what we’ve covered before [Bytesize#4](https://nf-co.re/events/2021/bytesize-4-github-contribution-basics). Git is a version control code system that has a bit of jargon that you need to know. You do `commit`, which are little packages of work, so you do some edits and then bundle them together and then say `commit` and that’s like checkpoints in the timeline of your code. They go in a kind of straight line, and then you can make branches, where you can work on different things at the same time.
```

[1:44](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=104) You can carry on working in that branch.

[1:47](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=108) And when you want to, you can merge that branch back into the main line of the commit history.

[1:55](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=115) You can fork an entire repository, which basically copies a repository of code into your account. So for example, I have forked an nf-core pipeline to my personal Github account.

[2:08](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=128) And you can carry on doing development on your fork. When you want to contribute what you’ve done back to the main repository within nf-core, you open up a `pull request` that is then merged into the main line of commits. Then everything you’ve worked on becomes a part of the main repository.

> [2:29](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=150) All our nf core pipelines have at least three branches called `dev`, `master` and `TEMPLATE`. The `master` branch is for default (after the first release), and should only ever have the latest stable release code. That is so that if someone runs Nextflow with the pipeline name, but does not specify a release, it can grab the default branch, which has the latest stable release.

Reviewer: ```suggestion
[2:29](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=150) All our nf-core pipelines have at least three branches called `dev`, `master` and `TEMPLATE`. The `master` branch is for default (after the first release), and should only ever have the latest stable release code. That is so that if someone runs Nextflow with the pipeline name, but does not specify a release, it can grab the default branch, which has the latest stable release.
```

[3:00](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=180) The `dev` branch is where we do all the development work, so things change there all the time. When we’re ready for a release, we merge `dev` into `master`.

[3:14](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=194) Finally, we have this kind of special branch called `TEMPLATE`, and that’s what I’m going to be talking about.

> [3:24](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=204) So this is one of the figures from the [nf-core paper](https://rdcu.be/clDKw) , and it describes how automated sync works from nf-core `TEMPLATE`. One of the main things we offer within the nf-core organisation is this pipeline template. You can do `nf-core create`, where it asks you a couple of questions i.e. a name, a description, an author, and it creates a minimal but functional pipeline. It has a lot of the boilerplate code that we think is best practice for a Nextflow pipeline. So it sets up lots of config stuff and it creates all the code structure into different files, it sets up automated tests and Github actions, continuous integration etc. But that `TEMPLATE` never stands still, we’re adding to it all the time, and it grows with the community and experience. So if we fix something in one of the pipelines, which is general, we try and bring that back into the main `TEMPLATE`. But of course once you’ve done `nf-core create`, your repository is then separate and no longer tracks along with the nf-core subject, and that’s what we’re trying to bring in here. So the trick we’ve figured out is that when we make your pipeline, you straightaway make your commits where you haven’t done any changes - it’s just a template. You have that initial `commit` with the `TEMPLATE`, and we branch at that point to a new branch called `TEMPLATE`, which has the same shared `commit`. You can then just forget about that `TEMPLATE` branch and carry on doing whatever you’re doing in `master` or `dev`, but when there’s a new release of the nf-core `TEMPLATE`, we create a new `commit` along this special `TEMPLATE` branch. Now we’ve got one `commit` in this to commit, and the only changes in this branch are the things we updated in the main `TEMPLATE`. This is always the case for the `TEMPLATE` branch, it only ever has the vanilla template from the main nf-core `TEMPLATE`, and the commits are only ever just updates from there as we do new releases of tools. When you get a new `commit`, then we can do a `pull request` from that, `branch` it into the `dev` branch so that anything that changed in the boilerplate code is brought into `dev`. But it never goes the other way, dev is never brought into `TEMPLATE`.

Reviewer: ```suggestion
[3:24](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=204) So this is one of the figures from the [nf-core paper](https://rdcu.be/clDKw), and it describes how automated sync works from nf-core `TEMPLATE`. One of the main things we offer within the nf-core organisation is this pipeline template. You can do `nf-core create`, where it asks you a couple of questions i.e. a name, a description, an author, and it creates a minimal but functional pipeline. It has a lot of the boilerplate code that we think is best practice for a Nextflow pipeline. So it sets up lots of config stuff and it creates all the code structure into different files, it sets up automated tests and Github actions, continuous integration etc. But that `TEMPLATE` never stands still, we’re adding to it all the time, and it grows with the community and experience. So if we fix something in one of the pipelines, which is general, we try and bring that back into the main `TEMPLATE`. But of course once you’ve done `nf-core create`, your repository is then separate and no longer tracks along with the nf-core template, and that’s what we’re trying to bring in here. So the trick we’ve figured out is that when we make your pipeline, you straightaway make your commits where you haven’t done any changes - it’s just a template. You have that initial `commit` with the `TEMPLATE`, and we branch at that point to a new branch called `TEMPLATE`, which has the same shared `commit`. You can then just forget about that `TEMPLATE` branch and carry on doing whatever you’re doing in `master` or `dev`, but when there’s a new release of the nf-core `TEMPLATE`, we create a new `commit` along this special `TEMPLATE` branch. Now we’ve got one `commit` in this to commit, and the only changes in this branch are the things we updated in the main `TEMPLATE`. This is always the case for the `TEMPLATE` branch, it only ever has the vanilla template from the main nf-core `TEMPLATE`, and the commits are only ever just updates from there as we do new releases of tools. When you get a new `commit`, then we can do a `pull request` from that, `branch` it into the `dev` branch so that anything that changed in the boilerplate code is brought into `dev`. But it never goes the other way, dev is never brought into `TEMPLATE`.
```

[7:33](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=453) We’ve got something special for you today, which is the `nf-core/tools` here. We’ve been working hard on a new release, and in fact it’s all prepped and ready to go.

[7:46](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=466) Here we are on the release page. I’ve just written the changelog and we’ve merged the master, all tests are passed, so I’m going to hit release now. We will see if the synchronisation process works (automated sync doesn’t work completely flawlessly all the way and will mostly fail).

[8:09](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=489) So I hit publish, and there you go, tools is released!

> [8:13](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=494) Now if I go to the Github access page, you can see lots of stuff is automatically happening here because of the release. The one that we are interested in is called `sync template`.

Reviewer: ```suggestion
[8:13](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=494) Now if I go to the Github actions page, you can see lots of stuff is automatically happening here because of the release. The one that we are interested in is called `sync template`.
```

[9:32](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=572) But you can see it committed changes and pushed to the `TEMPLATE` branch on the remote (found a new bug). It made a new branch, pushed it to remote and then tried to make an API and then found something wrong in here (a bug), and it’s crashed now, but it worked before it crashed.

[10:10](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=610) So thankfully, we have our automated `pull request`. Now if I go to this repository, we can have a look at the branches. You see the template branch here, and see a `commit` for each one of the releases of tools.

[10:25](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=625) This is the new commit that we just pushed, and these are all the things that have changed in the template since the last time.

[10:33](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=633) Now if I go to a special branch called `nf-core-template-merge`, it looks exactly the same and that’s where the `pull request` comes from.

> [10:42](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=642) So this is where you start if you’re a pipeline developer. You will just have got a `pull request` in your inbox that you will go in an see something like this that hopefully explains what to do.

Reviewer: ```suggestion
[10:42](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=642) So this is where you start if you’re a pipeline developer. You will just have got a `pull request` in your inbox that you will go in and see something like this that hopefully explains what to do.
```

[12:15](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=735) This looks more up-to-date and more like what we want to keep, so I delete this part of the merge conflict, and delete that marker and mark it as resolved. Then I just go on to the next one, and work through the rest. This is only really viable if the merge conflicts are few and simple because you need to manually deal with the merge conflict markers, and it’s quite easy to make a mistake. So generally, I would avoid doing this very often.

[12:44](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=764) Very often that button will be greyed out if there’s too much for Git to do on the web interface, and you will need to use the command line instead. You can basically do that on a branch, but I would recommend doing this on your fork instead.

[12:58](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=776) So the process would be to first update your personal fork at the repository against the `dev`, make a new branch to do this work, and then pull the changes from the upstream repository on the `TEMPLATE` branch.

> [13:17](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=797) If you watched [bytesize#11](https://nf-co.re/events/2021/bytesize-11-dev-envs-workflows) , I talked a little about how to put in changes from upstream repositories. So check that out if you’re not quite sure. But typically, you have two forks, you have origin which is your fork, and upstream which is the nf-core fork. So you can do ´git pull upstream template´ and that will pull the changes from the TEMPLATE branch (there will be a bunch of merge conflicts but you can fix those). You can commit them, push to your fork, and then open a pull request from your fork into the main repository.

Reviewer: ```suggestion
[13:17](https://youtu.be/-CZKoo5Y_08?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=797) If you watched [bytesize#11](https://nf-co.re/events/2021/bytesize-11-dev-envs-workflows), I talked a little about how to pull in changes from upstream repositories. So check that out if you’re not quite sure. But typically, you have two forks, you have origin which is your fork, and upstream which is the nf-core fork. So you can do ´git pull upstream template´ and that will pull the changes from the TEMPLATE branch (there will be a bunch of merge conflicts but you can fix those). You can commit them, push to your fork, and then open a pull request from your fork into the main repository.
```
